### ISSUE TITLE
Add support for rewriting using quantified equations

### ISSUE THREAD

**Issue by astump** [cedille/cedille#39]
Change the behavior of rho (and rho+) so that one can rewrite with a quantified equation, performing matching instead of just conversion-checking when trying to see if a side of the equation matches a subterm of the term/type/kind to be rewritten.

For example, one would be able to write

ρ addAssoc - t

and the correct instantiation (if any is possible) for the three quantified variables of our statement of associativity of addition would be automatically found.  This would be nicer than writing

ρ (addAssoc e1 e2 e3) - t

where e1, e2, and e3 are the instances for those variables.

addAssoc ◂ Π m : Nat. Π n : Nat. Π o : Nat. {add (add m n) o ≃ add m (add n o)}

**Comment by cwjnkins** [comment]
What should the behavior be if there are multiple such matches? E.g, if you had an expected type like

    {add (add (add e1 e2) e3) e4 ≃ t} -- for some terms e1..4 and t

Then the two possible matches are 

    -- 1) m = e1, n = e2, l = e3
    -- 2) m = (add e1 e2), n = e3, l = e4

Perhaps the default behavior should be to rewrite the outermost match? Or complain if multiple matches are possible (which could get hairy to detect if ρ+ is involved)

**Comment by astump** [comment]
Good point, Chris.  I think rewriting matches found in the current order of exploring the target term (top-down, left-to-right) would indeed be the right choice, certainly for extending the code we currently have in rewriting.agda.  Note that this should still allow rho to work with occurrence numbers (ρ<1 3 5> for the first, third, and fifth match), without any changes.
